packages/core/src/Modifiers.ts

```diff
@@ -247,6 +247,22 @@ export class Modifiers extends EventMixin {
     toggle(modifier: Modifier | Constructor<Modifier>): void {
         this.remove(modifier) || this.append(modifier);
     }
+    /**
+     * Check that all `otherModifiers` ar contained within this ones.
+     *
+     * @param otherModifiers
+     */
+    contains(otherModifiers: Modifiers): boolean {
```

Comment:
Du coup pour rester dans l'esprit des autres méthodes, ce serait bien d'aussi accepter `...modifiers: Array<Modifier | Constructor<Modifier>>`.

Reply:
I specifically need to check if `Modifiers` are contained, not `Modifier`.

Including `Constructor<Modifier>` would introduce a new logic that I think is useless. `Modifier.isSameAs` does not take `Constructor<Modifier>`.
If we need to check the constructor, then using a `.filter()` would be sufficient as we don't need to check the undefined.

On the other hand I could use `Modifier[]` in the signature. Which could have two forms:
```ts
    contains(modifiers: Modifiers | Modifier[]): boolean {
        //...
```
or:
```ts
    contains(...args: [Modifiers] | Modifier[]): boolean {
        const modifiers = args[0] instanceof Modifiers ? args[0] : (args as Modifier[]);
        //...
```
Which one you prefer @dmo-odoo @Zinston ?

I've push the first version as I found it more elegant but I'll change it if you have a different taste.

Reply:
The reason I suggested to include the other types was that most methods in `Modifiers` do and that this one is not fundamentally different from the others so from an API perspective I would expect to be able to use this one the same was as the others. Even though it admittedly adds a little bit of logic. As you said though:
> If we need to check the constructor, then using a .filter() would be sufficient as we don't need to check the undefined.
So there's little cost to putting that in `contains`.

Reply:
So what do I do?
Do I continue to take `Modifiers` as argument?
Do I add the logic for Contructor<Modifier> ?

If I take it all it means the function would become this:
```ts
    contains(...args: [Modifiers] | (Modifier | Constructor<Modifier>)[]): boolean {
        if (args[0] instanceof Modifiers || args[0] instanceof Modifier) {
            const modifiers =
                args[0] instanceof Modifiers
                    ? args[0]
                    : (args as (Modifier)[]);
            for (const modifier of modifiers) {
                const foundModifier = this.find(m => m.isSameAs(modifier));
                // Modifier.isSameAs(undefined) may return true in the case of
                // modifiers that are themselves containers of values. Then we might
                // want to equate the presence of an "empty" modifier with the
                // absence of that modifier.
                if (!foundModifier && !modifier.isSameAs(undefined)) {
                    return false;
                }
            } else {
                const modifierClasses = [...new Set(args as Constructor<Modifier>[])];
                return modifierClasses.every(this.find.bind(this))
            }
        }
        return true;
    }
```
Which becomes complicated to read for not so much value.
And also if I want consistencies, it means that all other method should have the same signature (i.e. including `Modifiers`).

Reply:
To summarize some of the options:
`...modifiers: (Modifier | Constructor<Modifier>)[]`
- I don't like because it does not take `Modifiers`
- add branching + logic for Contstructor<Modifier>

`modifiers: Modifiers | Modifier[]`
- +simple
- require to change the other signature for consistency
- add branching + logic for Contstructor<Modifier>

`modifiers: Modifiers | (Modifier | Constructor<Modifier>)[]`
- require to change the other signature for consistency
- add branching + logic for Contstructor<Modifier>

`...args: [Modifiers] | (Modifier | Constructor<Modifier>)[]`
- require to change the other signature for consistency
- add branching + logic for Contstructor<Modifier>

Reply:
And my view is that we should look at all those options individually but do the same as the other methods as that is what I would expect from an API.

Reply:
Mmh... Isn't it possible to make the `Modifiers` class compatible with `...` destructuring ?
Such that this would work:
```
function f(...modifiers: Modifier[] || Constructor<Modifier>[]) {}
```
```
const modifiers = new Modifiers();
f(...modifiers)
```
This way the `...` signature would work in all cases. It would work for `Modifier[]`, for `Constructor<Modifier>` *and* for `Modifiers` because `...` would transform it into `Modifier[]` through the iterable interface.

Reply:
I think the following would work:
```ts
const modifiers = new Modifiers();
f(...[modifiers]).
```

Reply:
The would be great because you could then give *any* iterable of modifiers to these functions, including `Modifiers` itself since it would be iterable.

Reply:
:+1:
[...modifiers] would magically call Modifiers.clone ?

Reply:
I've tested I it does not work. You need `f(...[...modifier])`.
Then the code would be:
```ts
    contains(...modifiers: Modifier[] | Constructor<Modifier>[]): boolean {
        if (modifiers[0] instanceof Modifier) {
            for (const modifier of modifiers as Modifier[]) {
                const foundModifier = this.find(m => m.isSameAs(modifier));
                // Modifier.isSameAs(undefined) may return true in the case of
                // modifiers that are themselves containers of values. Then we might
                // want to equate the presence of an "empty" modifier with the
                // absence of that modifier.
                if (!foundModifier && !modifier.isSameAs(undefined)) {
                    return false;
                }
            }
            return true;
        } else {
            const modifierClasses = [...new Set(modifiers as Constructor<Modifier>[])];
            return modifierClasses.every(this.find.bind(this));
        }
    }
    areSameAs(otherModifiers: Modifiers): boolean {
        return this.contains(...[...otherModifiers]) && otherModifiers.contains(...[...this]);
    }
```

Reply:
The first `...` is the syntax to specify that you want an array being considered as the list of arguments.
So you need to `...` over an array.
The second expression with `[...otherModifiers]` say take the iterator OtherModifier and place it in an array through the spreat iterator.
I think the first `...` is not considered as a spread operator although it is written the same way.

Reply:
Mmmh... Then what about this:
```
contains(...modifiers: Iterable<Modifier | Constructor<Modifier>): boolean {
```
Wouldn't `Modifiers` be an `Iterable<Modifier>` ?

Reply:
I already tried. I've tested in javascript and it would work.

The problem is typescript.
If you do `fn(...anything: AnyType) {}`, typescript tell you AnyType must be an array of something.
An iterable is not an Array so `fn(...anything: Iterable<any>) {}` cannot work.

And if you do `contains(...modifiers)` (which would work in js), typescript tells you:
> Argument of type 'unknown[]' is not assignable to parameter of type 'Modifier[] | Constructor<Modifier>[]'.